### PR TITLE
feat(graphql): Graphql Federaton for v1+v2

### PR DESCRIPTION
**Pull request by tomaszdurka**

### CODE REVIEW

File: libs/permissions/api/src/lib/permission.guard.ts [hunk 1]
```diff
@@ -24,6 +24,8 @@ export class PermissionsGuard implements CanActivate {
     const request = ctx.getContext().req
     const session = request.user
 
+    const {headers} = request
```
Comment (Itrulia) [inline libs/permissions/api/src/lib/permission.guard.ts:27]: I guess can be reverted

File: libs/settings/api/src/lib/settings.resolver.ts [hunk 2]
```diff
@@ -19,13 +19,13 @@ export class SettingsResolver {
   }
 
   @Query(returns => Setting, {
-    name: 'setting',
+    name: 'settingById',
```
Comment (Itrulia) [inline libs/settings/api/src/lib/settings.resolver.ts:22]: Can we revert this? As this is iirc inconsistent with the rest of the application.

File: libs/event/api/src/lib/event.model.ts [hunk 3]
```diff
@@ -13,12 +15,13 @@ import {Image} from '@wepublish/image/api'
...
 import {PaginatedType, SortOrder} from '@wepublish/utils/api'
 import {Node} from 'slate'
+import {Page} from './page.model'
 
 export enum EventSort {
-  CreatedAt = 'CreatedAt',
-  ModifiedAt = 'ModifiedAt',
-  StartsAt = 'StartsAt',
-  EndsAt = 'EndsAt'
+  CREATED_AT = 'CREATED_AT',
```
Comment (Itrulia) [inline libs/event/api/src/lib/event.model.ts:21]: Why? :(

**Comment by Itrulia** [comment]
Looking very good!

File: libs/event/api/src/lib/event.model.ts (hunk 3 above)
Comment (tomaszdurka) [reply libs/event/api/src/lib/event.model.ts:21]: There are differences between versions.
This is only thing left for me now in this pull-request.

E.g.

```
export enum SubscriptionDeactivationReason {
  InvoiceNotPaid = 'INVOICE_NOT_PAID',
  None = 'NONE',
  UserSelfDeactivated = 'USER_SELF_DEACTIVATED',
  InvoiceNotPaid = 'invoiceNotPaid',
  None = 'none',
  UserSelfDeactivated = 'userSelfDeactivated'
}
```

**Comment by tomaszdurka** [comment]
@Itrulia Please take a look again.

Currently server starts:
1. "publicApi" which serves `/v1/admin` and `/v1 gateway`
  - `/v1/admin` is actuall v1/admin apollo 
  - `/v1 gateway` is nestjs gateway combining v1 and v2

2. "privateApi" which server "v1" and "v2"
  - `/v1` is the actual v1 apollo, accessed only via `/v1 gateway`
  - `/v2` is the actual v2 nestjs

cc @elias-summermatter

File: libs/settings/api/src/lib/settings.resolver.ts (hunk 2 above)
Comment (tomaszdurka) [reply libs/settings/api/src/lib/settings.resolver.ts:22]: This is actually a conflict in graphql schema. v1 searches by name, v2 by id.

**Review by elias-summermatter — CHANGES_REQUESTED** [review]

File: libs/api/src/lib/graphql/blocks.ts [hunk 4]
```diff
@@ -1103,27 +1102,29 @@ export const GraphQLEventBlock = new GraphQLObjectType<EventBlock, Context>({
     filter: {type: new GraphQLNonNull(GraphQLEventBlockFilter)},
     events: {
       type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLEvent))),
-      resolve: ({filter}, _, {prisma}) =>
```
[the remark this answers is not in the document]
Comment (tomaszdurka) [reply libs/api/src/lib/graphql/blocks.ts:1106]: This was actually preparation of migration of events to v2. So then you don't rely on data form this, but only return ids with typename. Data is then retrieved via resolveReference on v2. 

However it can also coexist with types defined in v1 and v2 and both retrieving data on its own. Federation makes sure schema is unified.

File: apps/api-example/schema-v2.graphql [hunk 5]
```diff
@@ -2,6 +2,26 @@
 # THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
```
[the remark this answers is not in the document]
Comment (tomaszdurka) [reply apps/api-example/schema-v2.graphql:2]: 👀

**Comment by tomaszdurka** [comment]
This is heavily irritating. Tests are flaky and cannot make them reliably work.

File: package.json [hunk 6]
```diff
@@ -334,6 +337,8 @@
   },
   "optionalDependencies": {
     "@img/sharp-darwin-arm64": "^0.33.4",
-    "@img/sharp-linux-x64": "^0.33.4"
+    "@img/sharp-linux-x64": "^0.33.4",
+    "@nx/nx-darwin-arm64": "18.2.4",
```
Comment (Itrulia) [inline package.json:341]: Probably not needed anymore with the changes on master. This might cause your issues.

**Comment by tomaszdurka** [comment]
### Summary of changes current master /v1 vs current branch v1
```
Dangerous changes
  ⚠ Query.authors arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ Query.articles arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ Query.pages arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ Query.comments arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ Query.memberPlans arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ Query.events arg sort has changed defaultValue from STARTS_AT to StartsAt.
  ⚠ Query.events arg order has changed defaultValue from DESCENDING to Ascending.
  ⚠ An optional arg cursorId on Query.events was added.
  ⚠ Query.tags arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ Query.phrase arg order has changed defaultValue from DESCENDING to Descending.
  ⚠ An optional arg filter on Query.settings was added.
  ⚠ Cancelled was added to enum type EventStatus.
  ⚠ Rescheduled was added to enum type EventStatus.
  ⚠ Postponed was added to enum type EventStatus.
  ⚠ Scheduled was added to enum type EventStatus.
  ⚠ Ascending was added to enum type SortOrder.
  ⚠ Descending was added to enum type SortOrder.
  ⚠ monthly was added to enum type PaymentPeriodicity.
  ⚠ quarterly was added to enum type PaymentPeriodicity.
  ⚠ biannual was added to enum type PaymentPeriodicity.
  ⚠ yearly was added to enum type PaymentPeriodicity.
  ⚠ none was added to enum type SubscriptionDeactivationReason.
  ⚠ userSelfDeactivated was added to enum type SubscriptionDeactivationReason.
  ⚠ invoiceNotPaid was added to enum type SubscriptionDeactivationReason.
  ⚠ CreatedAt was added to enum type EventSort.
  ⚠ ModifiedAt was added to enum type EventSort.
  ⚠ StartsAt was added to enum type EventSort.
  ⚠ EndsAt was added to enum type EventSort.
  ⚠ MAIL_PROVIDER_NAME was added to enum type SettingName.

BREAKING CHANGES
  ✖ Point was removed.
  ✖ EventConnection was removed.
  ✖ Value was removed.
  ✖ Query.events arg cursor was removed.
  ✖ Query.events changed type from EventConnection to PaginatedEvents!.
  ✖ Image.focalPoint changed type from Point to FocalPoint.
  ✖ Event.tags changed type from [Tag!]! to [Tag!].
  ✖ CANCELLED was removed from enum type EventStatus.
  ✖ RESCHEDULED was removed from enum type EventStatus.
  ✖ POSTPONED was removed from enum type EventStatus.
  ✖ SCHEDULED was removed from enum type EventStatus.
  ✖ ASCENDING was removed from enum type SortOrder.
  ✖ DESCENDING was removed from enum type SortOrder.
  ✖ MONTHLY was removed from enum type PaymentPeriodicity.
  ✖ QUARTERLY was removed from enum type PaymentPeriodicity.
  ✖ BIANNUAL was removed from enum type PaymentPeriodicity.
  ✖ YEARLY was removed from enum type PaymentPeriodicity.
  ✖ NONE was removed from enum type SubscriptionDeactivationReason.
  ✖ USER_SELF_DEACTIVATED was removed from enum type SubscriptionDeactivationReason.
  ✖ INVOICE_NOT_PAID was removed from enum type SubscriptionDeactivationReason.
  ✖ EventFilter.tags changed type from [ID!] to [String!].
  ✖ STARTS_AT was removed from enum type EventSort.
  ✖ ENDS_AT was removed from enum type EventSort.
  ✖ CREATED_AT was removed from enum type EventSort.
  ✖ MODIFIED_AT was removed from enum type EventSort.
  ✖ Setting.value changed type from Value! to GraphQLSettingValueType.
  ✖ PollVote.answerId changed type from String! to ID!.
    ```

**Comment by tomaszdurka** [comment]
### Summary of changes current master /v2 vs current branch /v1

https://gist.github.com/tomaszdurka/c7b716b9ce9024a0422cbf0aacbf279e

```
BREAKING CHANGES
  ✖ MemberPlanRef was removed.
  ✖ PaymentMethodRef was removed.
  ✖ Event.id changed type from String! to ID!.
  ✖ EventFromSource.id changed type from String! to ID!.
  ✖ Image.description changed type from RichText to String.
  ✖ Image.id changed type from String! to ID!.
  ✖ PollVote.disabled changed type from String! to Boolean!.
  ✖ Query.image was removed.
  ✖ Query.settingsList was removed.
  ✖ Query.event arg id has changed type from String! to ID!.
  ✖ Query.events arg cursorId has changed type from String to ID.
  ✖ Query.paymentMethods changed type from [PaymentMethodRef!]! to [PaymentMethod!]!.
  ✖ Query.setting arg id was removed.
  ✖ A required arg name on Query.setting was added.
  ✖ SubscriptionFlowModel.memberPlan changed type from MemberPlanRef to MemberPlan.
  ✖ SubscriptionFlowModel.paymentMethods changed type from [PaymentMethodRef!]! to [PaymentMethod!]!.
  ✖ User.active was removed.
  ✖ User.createdAt was removed.
  ✖ User.modifiedAt was removed.
  ✖ User.password was removed.
  ✖ User.roleIDs changed type from [String!]! to [String!].
```

File: apps/api-example/src/nestapp/gateway.module.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import {ApolloGatewayDriver, ApolloGatewayDriverConfig} from '@nestjs/apollo'
+import {Module} from '@nestjs/common'
+import {GraphQLModule} from '@nestjs/graphql'
+
+@Module({
+  imports: [
+    GraphQLModule.forRoot<ApolloGatewayDriverConfig>({
+      driver: ApolloGatewayDriver,
+      server: {
+        introspection: true,
```
Comment (elias-summermatter) [inline apps/api-example/src/nestapp/gateway.module.ts:11]: We should use the apolloIntrospection field from the config file. If thats easy possible if not thats fine with me too.

File: apps/api-example/src/nestapp/gateway.module.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        sortSchema: true,
+        path: 'v1',
+        cache: 'bounded',
+        playground: process.env.NODE_ENV === 'development',
```
Comment (elias-summermatter) [inline apps/api-example/src/nestapp/gateway.module.ts:15]: Here also use the filed from the config "apolloPlayground"

File: package.json [hunk 9]
```diff
@@ -336,6 +339,22 @@
   },
   "optionalDependencies": {
     "@img/sharp-darwin-arm64": "^0.33.4",
-    "@img/sharp-linux-x64": "^0.33.4"
+    "@img/sharp-linux-x64": "^0.33.4",
```
Comment (elias-summermatter) [inline package.json:342]: Is that correct that all of these extra dependencies need to be added?

File: package.json [hunk 10]
```diff
@@ -20,7 +20,7 @@
     "watch:api-example": "nx serve api-example",
     "watch:website-example": "nx serve website-example",
     "watch:media": "nx serve media",
-    "start:docker": "docker-compose up --detach database pgadmin storage",
```
Comment (elias-summermatter) [inline package.json:23]: Why is this removed?

File: Dockerfile [hunk 11]
```diff
@@ -47,7 +47,7 @@ RUN apt-get update && \
     apt-get install -y --no-install-recommends openssl && \
     npm ci && \
     npm install -g pkg && \
-    npx nx build api-example && \
+    npx nx build api-example --skip-nx-cache && \
```
[the remark this answers is not in the document]
Comment (elias-summermatter) [reply Dockerfile:50]: @tomaszdurka what about this?

**Review by elias-summermatter — CHANGES_REQUESTED** [review]

File: apps/api-example/src/nestapp/gateway.module.ts (hunk 8 above)
Comment (tomaszdurka) [reply apps/api-example/src/nestapp/gateway.module.ts:15]: Oh I've tried, but I cannot use `GraphQLModule.forRootAsync` seems it's not accepting driver property.... strange. Probably bug with graphql/nestjs.

Comment (tomaszdurka) [reply apps/api-example/src/nestapp/gateway.module.ts:15]: Nvm, actually I can. Just strange driver was put outside of the useFactory returned options.
